File: sorting/quicksort.c
#include <stdio.h>


void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int* container, int left, int right, int pivot){
    while(left <= right){
        while(container[left++] < pivot){}
        while(container[right--] > pivot){}
        if(left <= right){
            printf("%d, %d\n", container[left], container[right]);
            swap(&container[left], &container[right]);
            printf("%d, %d\n", container[left], container[right]);
            left++;
            right--;
        }
    }
    return left;
}

void qs(int* container, int left, int right){
    if(left >= right){ return; }
    int middle = (left + right) / 2;
    int pivot = container[middle];
    int index = partition(container, left, right, pivot);
    qs(container, left, index - 1);
    qs(container, index, right);
}

int main(int argc, const char* argv[]){
    int array[] = {1, 32, 12, 19};
    qs(array, 0, 4);
    for(int i = 0; i < 4; ++i){
        printf("%d ", array[i]);
    }
    printf("\n");
    return 0;
}

File: sorting/quicksort.py
#!/usr/bin/env python3.8

import random

def qs(container: list, left: int, right: int):
    if(left >= right):
        return
    pivot = container[int((left + right)/2)]
    index = partition(container, left, right, pivot)
    qs(container, left, index - 1)
    qs(container, index, right)

def partition(container: list, left: int, right: int, pivot: int):
    while(left <= right):
        while(container[left] < pivot):
            left+=1
        while(container[right] > pivot):
            right-=1
        if(left <= right):
            container[left], container[right] = container[right], container[left]
            left+=1
            right-=1
    return left

def func(container: list):
    qs(container, 0, len(container)-1)

def test():
    example = [random.randint(0, 1000) for x in range(0, 100)]
    example_copy = example
    func(example)
    return (example == sorted(example_copy))

for x in range(0, 1000):
    if not(test()):
        print(f'failed at test {x}')
        quit()
print("finished all tests!")

